Source/HTTP/XCurl/XCurlMulti.cpp

```diff
@@ -0,0 +1,154 @@
+    }
+}
+
+HRESULT XCurlMulti::AddRequest(HC_UNIQUE_PTR<XCurlEasyRequest>&& easyRequest)
+{
+    std::unique_lock<std::mutex> lock{ m_mutex };
+
+    auto result = curl_multi_add_handle(m_curlMultiHandle, easyRequest->Handle());
+    if (result != CURLM_OK)
+    {
+        HC_TRACE_ERROR(HTTPCLIENT, "XCurlMulti::AddRequest: curl_multi_add_handle failed with CURLCode=%u", result);
+        return HrFromCurlm(result);
+    }
+
+    m_easyRequests.emplace(easyRequest->Handle(), std::move(easyRequest));
+
+    RETURN_IF_FAILED(XTaskQueueSubmitCallback(m_queue, XTaskQueuePort::Work, this, XCurlMulti::TaskQueueCallback));
+
+    return S_OK;
+}
+
+void CALLBACK XCurlMulti::TaskQueueCallback(_In_opt_ void* context, _In_ bool canceled) noexcept
+{
+    if (!canceled)
+    {
+        auto multi = static_cast<XCurlMulti*>(context);
+        HRESULT hr = multi->Perform();
+        if (FAILED(hr))
+        {
+            HC_TRACE_ERROR_HR(HTTPCLIENT, hr, "XCurlMulti::Perform failed. Failing all active requests.");
+            multi->FailAllRequests(hr);
+        }
+    }
+}
+
+HRESULT XCurlMulti::Perform() noexcept
+{
+    std::unique_lock<std::mutex> lock{ m_mutex };
+
+    int runningRequests{ 0 };
+    CURLMcode result = curl_multi_perform(m_curlMultiHandle, &runningRequests);
+    if (result != CURLM_OK)
+    {
+        HC_TRACE_ERROR(HTTPCLIENT, "XCurlMulti::Perform: curl_multi_perform failed with CURLCode=%u", result);
+        return HrFromCurlm(result);
+    }
+
+    int remainingMessages{ 1 }; // assume there is at least 1 message so loop is always entered
+    while (remainingMessages)
+    {
+        CURLMsg* message = curl_multi_info_read(m_curlMultiHandle, &remainingMessages);
+        if (message)
+        {
+            switch (message->msg)
+            {
+            case CURLMSG_DONE:
+            {
+                auto requestIter = m_easyRequests.find(message->easy_handle);
+                assert(requestIter != m_easyRequests.end());
+
+                result = curl_multi_remove_handle(m_curlMultiHandle, message->easy_handle);
+                if (result != CURLM_OK)
+                {
+                    HC_TRACE_ERROR(HTTPCLIENT, "XCurlMulti::Perform: curl_multi_remove_handle failed with CURLCode=%u", result);
+                }
+
+                requestIter->second->Complete(message->data.result);
+                m_easyRequests.erase(requestIter);
+            }
+            break;
+            case CURLMSG_NONE:
+            case CURLMSG_LAST:
+            default:
+            {
+                HC_TRACE_ERROR(HTTPCLIENT, "XCurlMulti::Perform: Unrecognized CURLMsg!");
+                assert(false);
+            }
+            break;
+            }
+        }
+    }
+
+    if (runningRequests)
+    {
+        // Reschedule Perform if there are still running requests
+        RETURN_IF_FAILED(XTaskQueueSubmitCallback(m_queue, XTaskQueuePort::Work, this, XCurlMulti::TaskQueueCallback));
```

Comment:
is this going to be a "busy loop" as long as a request is outstanding?

what happens if a game tries to empty the work queue while we're making an http request that takes 2s to complete?

Reply:
It will get continually scheduled, yes. A couple thoughts/questsions:

1. can a client empty/terminate a TaskQueue port?  XCurlMulti is using a composite TaskQueue that it creates.  My understanding is clients can terminate their queues without this being affected.

2. Should we be rescheduling the next Perform with a delay?  I originally intended to do that but XCurl doesn't support curl_multi_timeout ([https://curl.se/libcurl/c/curl_multi_timeout.html](https://curl.se/libcurl/c/curl_multi_timeout.html)).  We could just use a small fixed, delay though

Reply:
so consider a game that does this:
```
// inside some game loop (main or worker thread)
while (XAsyncDoWork(myQueue, false)) {}
do_other_important_per_frame_work();
```

Your approach will stall this thread until the http request is completed

Ideally we would have a way to only schedule work if there's something to be done (a callback version of `curl_multi_timeout`) since we don't have that, I think scheduling with a delay is the right way to go

Reply:
Ya, from what I can tell, there isn't a way to get a callback from curl when there is perform work to be done. I will add a small delay between each Perform